File: src/lib/channels.ts
import { invoke } from "@tauri-apps/api/core";

import { isTauriAvailable } from "./tauri";
import type {
  ChannelActivityEntry,
  ChannelDetail,
  ChannelSummary,
  ChannelUpsertInput,
  TelegramBotValidation,
  TelegramChatCandidate,
} from "../types";

const CHANNEL_STORAGE_KEY = "orchestra.mock.channels";
const CHANNEL_ACTIVITY_STORAGE_KEY = "orchestra.mock.channel-activity";

export function normalizeOptionalString(value?: string | null) {
  const trimmed = value?.trim();
  return trimmed ? trimmed : null;
}

function nowIso() {
  return new Date().toISOString();
}

function createId(prefix: string) {
  return `${prefix}-${Math.random().toString(36).slice(2, 8)}-${Date.now().toString(36)}`;
}

function getStoredChannels() {
  const value = window.localStorage.getItem(CHANNEL_STORAGE_KEY);
  return value ? (JSON.parse(value) as ChannelDetail[]) : [];
}

function saveStoredChannels(channels: ChannelDetail[]) {
  window.localStorage.setItem(CHANNEL_STORAGE_KEY, JSON.stringify(channels));
}

function getStoredChannelActivity() {
  const value = window.localStorage.getItem(CHANNEL_ACTIVITY_STORAGE_KEY);
  return value ? (JSON.parse(value) as ChannelActivityEntry[]) : [];
}

function saveStoredChannelActivity(entries: ChannelActivityEntry[]) {
  window.localStorage.setItem(CHANNEL_ACTIVITY_STORAGE_KEY, JSON.stringify(entries));
}

export async function listChannels(): Promise<ChannelSummary[]> {
  if (!isTauriAvailable()) {
    return getStoredChannels();
  }
  return invoke<ChannelSummary[]>("list_channels");
}

export async function getChannel(channelId: string): Promise<ChannelDetail> {
  if (!isTauriAvailable()) {
    const channel = getStoredChannels().find((entry) => entry.id === channelId);
    if (!channel) {
      throw new Error(`Channel ${channelId} was not found.`);
    }
    return channel;
  }
  return invoke<ChannelDetail>("get_channel", { channelId });
}

export async function listChannelActivity(channelId: string, limit = 50): Promise<ChannelActivityEntry[]> {
  if (!isTauriAvailable()) {
    return getStoredChannelActivity().filter((entry) => entry.channelId === channelId).slice(0, limit);
  }
  return invoke<ChannelActivityEntry[]>("list_channel_activity", { channelId, limit });
}

export async function createChannel(input: ChannelUpsertInput): Promise<ChannelDetail> {
  if (!isTauriAvailable()) {
    const timestamp = nowIso();
    const channel: ChannelDetail = {
      id: createId("channel"),
      kind: input.kind ?? "telegram",
      name: input.name?.trim() || "Telegram",
      enabled: Boolean(input.enabled),
      status: input.telegram?.botToken && input.telegram?.chatId ? "ready" : "needs_setup",
      targetAgentId: input.targetAgentId ?? "agent-supervisor",
      defaultProjectId: input.defaultProjectId ?? "orchestra",
      defaultProjectName: input.defaultProjectId ?? "orchestra",
      secretConfigured: Boolean(input.telegram?.botToken),
      telegram: {
        botUsername: null,
        apiBaseUrl: normalizeOptionalString(input.telegram?.apiBaseUrl),
        chatId: input.telegram?.chatId ?? null,
        chatTitle: input.telegram?.chatTitle ?? null,
        chatType: input.telegram?.chatType ?? null,
        commandsEnabled: input.telegram?.commandsEnabled ?? true,
      },
      lastError: null,
      lastActivityAt: null,
      createdAt: timestamp,
      updatedAt: timestamp,
    };
    saveStoredChannels([channel, ...getStoredChannels()]);
    return channel;
  }
  return invoke<ChannelDetail>("create_channel", { input });
}

export async function updateChannel(channelId: string, input: ChannelUpsertInput): Promise<ChannelDetail> {
  if (!isTauriAvailable()) {
    const channels = getStoredChannels();
    const existing = channels.find((entry) => entry.id === channelId);
    if (!existing) {
      throw new Error(`Channel ${channelId} was not found.`);
    }
    const updated: ChannelDetail = {
      ...existing,
      name: input.name?.trim() || existing.name,
      enabled: input.enabled ?? existing.enabled,
      defaultProjectId: input.defaultProjectId ?? existing.defaultProjectId,
      secretConfigured: existing.secretConfigured || Boolean(input.telegram?.botToken),
      telegram: {
        ...(existing.telegram ?? {
          botUsername: null,
          apiBaseUrl: null,
          chatId: null,
          chatTitle: null,
          chatType: null,
          commandsEnabled: true,
        }),
        apiBaseUrl: normalizeOptionalString(input.telegram?.apiBaseUrl) ?? existing.telegram?.apiBaseUrl ?? null,
        chatId: input.telegram?.chatId ?? existing.telegram?.chatId ?? null,
        chatTitle: input.telegram?.chatTitle ?? existing.telegram?.chatTitle ?? null,
        chatType: input.telegram?.chatType ?? existing.telegram?.chatType ?? null,
        commandsEnabled: input.telegram?.commandsEnabled ?? existing.telegram?.commandsEnabled ?? true,
      },
      status: (existing.secretConfigured || Boolean(input.telegram?.botToken)) && (input.telegram?.chatId ?? existing.telegram?.chatId)
        ? "ready"
        : "needs_setup",
      updatedAt: nowIso(),
    };
    saveStoredChannels(channels.map((entry) => (entry.id === channelId ? updated : entry)));
    return updated;
  }
  return invoke<ChannelDetail>("update_channel", { channelId, input });
}

export async function deleteChannel(channelId: string): Promise<void> {
  if (!isTauriAvailable()) {
    saveStoredChannels(getStoredChannels().filter((entry) => entry.id !== channelId));
    saveStoredChannelActivity(getStoredChannelActivity().filter((entry) => entry.channelId !== channelId));
    return;
  }
  await invoke("delete_channel", { channelId });
}

export async function validateTelegramBot(botToken: string, apiBaseUrl?: string | null): Promise<TelegramBotValidation> {
  if (!isTauriAvailable()) {
    return {
      botId: "mock-bot",
      username: "mock_orchestra_bot",
      displayName: "Mock Orchestra Bot",
    };
  }
  return invoke<TelegramBotValidation>("validate_telegram_bot", {
    botToken,
    apiBaseUrl: normalizeOptionalString(apiBaseUrl),
  });
}

export async function listTelegramChatCandidates(botToken: string, apiBaseUrl?: string | null): Promise<TelegramChatCandidate[]> {
  if (!isTauriAvailable()) {
    return [
      {
        chatId: "mock-chat",
        title: "Mock Telegram Chat",
        chatType: "private",
        username: "mock-user",
        lastMessageText: "/start",
        lastMessageAt: nowIso(),
      },
    ];
  }
  return invoke<TelegramChatCandidate[]>("list_telegram_chat_candidates", {
    botToken,
    apiBaseUrl: normalizeOptionalString(apiBaseUrl),
  });
}

File: src-tauri/src/commands/channels.rs
use tauri::{AppHandle, State};

use crate::{
    models::{
        ChannelActivityEntry, ChannelDetail, ChannelSummary, ChannelUpsertInput,
        TelegramBotValidation, TelegramChatCandidate,
    },
    services::{channels, database},
    state::AppState,
};

#[tauri::command]
pub fn list_channels() -> Result<Vec<ChannelSummary>, String> {
    let connection = database::open_connection()?;
    channels::list_channels(&connection)
}

#[tauri::command]
pub fn get_channel(channel_id: String) -> Result<ChannelDetail, String> {
    let connection = database::open_connection()?;
    channels::get_channel(&connection, &channel_id)
}

#[tauri::command]
pub fn list_channel_activity(
    channel_id: String,
    limit: Option<usize>,
) -> Result<Vec<ChannelActivityEntry>, String> {
    let connection = database::open_connection()?;
    channels::list_channel_activity(&connection, &channel_id, limit.unwrap_or(50))
}

#[tauri::command]
pub fn create_channel(
    app: AppHandle,
    state: State<'_, AppState>,
    input: ChannelUpsertInput,
) -> Result<ChannelDetail, String> {
    let connection = database::open_connection()?;
    let detail = channels::create_channel(&connection, input)?;
    channels::sync_channel_runtimes(app, &state)?;
    Ok(detail)
}

#[tauri::command]
pub fn update_channel(
    app: AppHandle,
    state: State<'_, AppState>,
    channel_id: String,
    input: ChannelUpsertInput,
) -> Result<ChannelDetail, String> {
    let connection = database::open_connection()?;
    let detail = channels::update_channel(&connection, &channel_id, input)?;
    channels::sync_channel_runtimes(app, &state)?;
    Ok(detail)
}

#[tauri::command]
pub fn delete_channel(
    app: AppHandle,
    state: State<'_, AppState>,
    channel_id: String,
) -> Result<(), String> {
    let connection = database::open_connection()?;
    channels::delete_channel(&connection, &channel_id)?;
    channels::sync_channel_runtimes(app, &state)?;
    Ok(())
}

#[tauri::command]
pub fn validate_telegram_bot(
    state: State<'_, AppState>,
    bot_token: String,
    api_base_url: Option<String>,
) -> Result<TelegramBotValidation, String> {
    match channels::validate_telegram_bot(&bot_token, api_base_url.as_deref()) {
        Ok(validation) => Ok(validation),
        Err(error) => {
            state.log("error", "channels.telegram.validate_bot.failed", &error);
            Err(error)
        }
    }
}

#[tauri::command]
pub fn list_telegram_chat_candidates(
    state: State<'_, AppState>,
    bot_token: String,
    api_base_url: Option<String>,
) -> Result<Vec<TelegramChatCandidate>, String> {
    match channels::list_telegram_chat_candidates(&bot_token, api_base_url.as_deref()) {
        Ok(candidates) => Ok(candidates),
        Err(error) => {
            state.log("error", "channels.telegram.list_chat_candidates.failed", &error);
            Err(error)
        }
    }
}
